### DISCUSSION TITLE
API Document Creation Breaks Document Management

### DISCUSSION THREAD

**Discussion by Amzamiviram** [open-webui/open-webui#4692] [General]
# Bug Report

## Installation Method

Docker running on Ubuntu 22.04

## Environment

- **Open WebUI Version:** v0.3.13 on docker running on Ubuntu 22.04
- **Ollama:** v0.3.6 Windows Preview running on Windows 11

**Confirmation:**

- [+ ] I have read and followed all the instructions provided in the README.md.
- [+] I am on the latest version of both Open WebUI and Ollama.
- [+] I have included the browser console logs.
- [+] I have included the Docker container logs.
- [+] I have provided the exact steps to reproduce the bug in the "Steps to Reproduce" section below.

## Expected Behavior:

I create a document using the open-webui API.  

## Actual Behavior:

Server returns "Internal Server Error", upon my request then documents cannot not be retrieved with the Get documents API(/api/v1/documents/).  The Get command also returns "Internal Server Error" and the website interface /workspace/documents page no longer can be loaded.

## Description

**Bug Summary:**
Creating a document via the API seems to break the ability to view/upload/delete documents both in the web interface and API

## Reproduction Details

**Steps to Reproduce:**
1. Create the docker container, create an account, and get an auth token from the API.

2. Create any document via the API, returns result "Internal Server Error".

From Powershell, the command being used is

`Invoke-RestMethod -Method Post -Uri 'http://192.168.100.10:3000/api/v1/documents/create' -ContentType 'application/json' -Authentication Bearer -Token $Token -Body $body  -AllowUnencryptedAuthentication`

where $body is 
`{
  "name": "test5",
  "title": "test4",
  "collection_name": "test3",
  "filename": "test",
  "content": "test2"
}`

3. Now it is not possible to upload/view/delete any documents from either the API or web interface, both return "Internal Server Error" when attempting to get, remove, or add additional documents.

## Logs and Screenshots

**Browser Console Logs:**
[console-export-2024-8-17_12-39-5.txt](https://github.com/user-attachments/files/16646317/console-export-2024-8-17_12-39-5.txt)

**Docker Container Logs:**
[Docker Container.log](https://github.com/user-attachments/files/16646302/Docker.Container.log)

**Comment by hasani114** [comment]
Where are you getting the api/v1/documenta/create endpoint? The documentation I was able to access at http://localhost:3000/rag/api/v1/docs#/ does not have that endpoint at all.

**Reply by Amzamiviram** [reply]
There's multiple APIs available,  there's the rag one you mention as well as:
http://localhost:3000/images/api/v1/docs#/
http://localhost:3000/openai/docs#/
http://localhost:3000/ollama/docs
http://localhost:3000/api/v1/docs

this is described at the bottom of the page, under swagger documentation
https://docs.openwebui.com/features/

**Reply by hasani114** [reply]
I was on a previous version of open-webui and was able to reproduce this. I want to work on the document management to auto-sync my library every day so tried it out. Now I can't seem to undo this issue even recreated the container. How were you able to fix it to get it to a usable state? or is that not possible without losing my entire data folder?

**Reply by Amzamiviram** [reply]
I'm not sure how to fix it, in my case I don't have that much stored in open-webui so I just backed up my tools/functions and recreated the container.  I was trying the API to upload documents because using the "Scan for documents from DOCS_DIR (/data/docs)" from the admin panel only pulled in about a third of the documents I put there.

**Reply by hasani114** [reply]
I did recreate the container but not the entire data folder. I did delete the entire vector_db folder to see if that works, but nope. I'll have to back up configs and redo everything. 

The documents management needs a lot of improvement. I saw that if a document is already in the vector store scan/syncing doesn't reupload them if there are updates / changes which is a bummer. I'd have liked if there was an option to resync based on updates or even resync them entirely (since the cost for embeddings is so low) on a specific schedule.

**Reply by antonio-lamberti** [reply]
@hasani114 @Amzamiviram Did anyone succeed in uploading files via API? I also want to do a daily files update...But like in your case, the documents section broke! 
Did you find a solution in the end?

**Comment by GeneralNumpty** [comment]
For the /api/v1/documents/create API either pass in null if you don't want to add tags or to add tags they must be a json string which is stored in an SQLite table in the back end so you have to escape the double quotes in your response body like.

`"content": "{\"tags\":[{\"name\": \"newtag01\"}]}"`
or
`"content": null`

so using the http://localhost/api/v1/documents/create API to add a document collection with three tags we would have a request body like this:

```
{
"name": "temp_file01pdf",
"title": "temp_file-01.pdf",
"collection_name": "this is a unique collection name",
"filename": "temp_file-01.pdf",
"content": "{\"tags\":[{\"name\": \"newtag01\"},{\"name\": \"newtag02\"},{\"name\": \"newtag03\"}]}"
}
```

Notes
name field can be any unique 256 length string, the UI seems to take the file name and remove any special chars except underscores so temp_file-01.pdf becomes temp_file01pdf.
collection_name field is also a unique 256 length string, the WebUI seems to generates a UUID by default (presumably to stop name collisions).

There is limited data validation on the API so if you enter the wrong data particularly on the tags field it can hang the documents tab in the webUI.

To fix this there is a SQLlite DB that holds the config (in my Ubuntu setup its located in something like myhome/myenvironment/lib/python3.11/site-packages/open_webui/data/webui.db (I found this my monitoring the UI process for file access), just edit the DB (I used sqlitebrowser) to manually delete rows that have the problem JSON tag content (under the documents table) remember to write the DB changes before reopening the documents page.
